js-packages/@prestojs/viewmodel/src/ViewModelCache.ts

```diff
@@ -9,10 +9,49 @@ import ViewModel, { PrimaryKey } from './ViewModel';
  * to just check if it's an instanceof RecordPointer
  */
 class RecordPointer<T extends ViewModel> {
+    /**
+     * This is the value that was replaced by this pointer in the cache. When we actually clone
+     * the record we compare against this value - if it's the same we return the original object
+     * so that equality checks still hold. This is because we never return the RecordPointer - we
+     * only resolve the underlying value as is required. In order to determine whether or not
+     * an object is the same requires first cloning it and then comparing it. As such we do that
+     * at the time we need the cloned value, not upfront (ie. we don't do it before we store the
+     * RecordPointer as that would be lower than doing it lazily - especially if you are updating
+     * lots of records)
```

Comment:
Why `currentValue` rather than `currentRecord`?

I'm confused by the explanation. Up to line 15 I understand; it boils down to:

"This is the existing record. We only replace the record with a new one if field values actually change (improves performance by allowing object equality checks to do an object identity check first before having to do a deep equality check, potentially saving many rerenders)".

Everything after that is confusing (eg "we only resolve the underlying value as required") -- can we chat about it?

Reply:
What about:

> This is the existing record. We only replace a record with a new one if the fields actually change and we
> only do this when the record is accessed (if it's never accessed there's no performance and minimal
> storage costs for it).  When the record is accessed and it is a `RecordPointer` we have to clone it so
> the underlying record is returned. As part of this clone we compare `currentRecord` to the cloned
> version; if they are the same we discard the clone and return `currentRecord`.  This is purely a
> performance optimisation.

Also in case it's not clear the whole reason `RecordPointer` exists is for when you have update a set of fields (`email`, `name`, `phone`) then when records with a subset of fields are requested they get the updated values (without having to create everything upfront).